core/src/main/java/feign/FeignException.java

```diff
@@ -42,27 +44,65 @@ public int status() {
     return this.status;
   }
 
+
+  public String body() {
+    return body;
+  }
+
   static FeignException errorReading(Request request, Response ignored, IOException cause) {
-    return new FeignException(
-        format("%s reading %s %s", cause.getMessage(), request.method(), request.url()),
-        cause);
+    String message = format("%s reading %s %s", cause.getMessage(), request.method(), request.url());
+    
+    FeignException exc = new FeignException(message, cause);
+    exc.body = fetchRequestBody(request);
+
+    return exc;
   }
 
+
   public static FeignException errorStatus(String methodKey, Response response) {
     String message = format("status %s reading %s", response.status(), methodKey);
+
+    FeignException exception = new FeignException(response.status(), message);
+    exception.body = fetchResponseBody(response);
+
+    return exception;
+  }
+
+  static FeignException errorExecuting(Request request, IOException cause) {
+    String message = format("%s executing %s %s", cause.getMessage(), request.method(), request.url());
+    
+    RetryableException exception = new RetryableException(message, cause,null);
+    exception.body = fetchRequestBody(request);
+
+    return exception;
+  }
+
+  private static String fetchRequestBody(Request request) {
+    if (request.body() != null) {
+      try {
+        // no way of getting body's charset, assuming UTF-8, other charsets may generate
+        // errors, hence the catch and ignore
```

Comment:
You can get the charset from the Accept/Content-Type headers, but it isn't guaranteed to be there (there is also a `Charset.defaultCharset()` method). However, why not just store it as a `byte[]` as mentioned above (rather than taking the `byte[]` and applying a possibly-incorrect charset to it)? That sounds like the cleaner solution.